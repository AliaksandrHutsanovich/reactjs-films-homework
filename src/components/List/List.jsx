import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { List as AntdList } from 'antd';
import clsx from 'clsx';
import Text from '../Text';

import '!style-loader!css-loader!sass-loader!antd/dist/antd.css'; //eslint-disable-line

import styles from './List.scss';

const { Item } = AntdList;

export const renderListItem = (item) => (
  <Item className={styles.listItem}>
    <Text type="white" size={12}>
      {item.name}
    </Text>
  </Item>
);

const List = ({
  dataSource,
  grid,
  renderItem,
  bordered,
  rowKey,
  size,
  itemAutoWidth,
  itemMarginLeft,
}) => (
  <AntdList
    dataSource={dataSource}
    grid={grid}
    renderItem={renderItem}
    bordered={bordered}
    rowKey={rowKey}
    size={size}
    className={clsx({
      [styles.list]: itemAutoWidth,
      [styles[`itemMarginLeft_${itemMarginLeft}`]]: itemMarginLeft,
    })}
  />
);

List.propTypes = {
  dataSource: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.shape({}),
      PropTypes.node,
    ]),
  ).isRequired,
  grid: PropTypes.shape({
    gutter: PropTypes.number,
    column: PropTypes.number,
    xs: PropTypes.number,
    sm: PropTypes.number,
    md: PropTypes.number,
    lg: PropTypes.number,
    xl: PropTypes.number,
  }),
  renderItem: PropTypes.func,
  bordered: PropTypes.bool,
  rowKey: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
  ]),
  size: PropTypes.oneOf([
    'default',
    'large',
    'small',
  ]),
  itemAutoWidth: PropTypes.bool,
  itemMarginLeft: PropTypes.oneOf([
    5, 10, 15, 20,
  ]),
};

List.defaultProps = {
  grid: null,
  renderItem: renderListItem,
  bordered: false,
  rowKey: 'id',
  size: 'default',
  itemAutoWidth: false,
  itemMarginLeft: null,
};

export default memo(List); // writting memozator is possible
