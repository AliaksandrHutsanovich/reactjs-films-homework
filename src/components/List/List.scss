.listItem {
  padding-left: 0px !important;
  padding-right: 0px !important;
  margin-bottom: 2px !important;
}

.list > div > div > div > :nth-child(n) {
  width: auto !important;
  max-width: none !important;
}

@for $var from 1 to 4 {
  .itemMarginLeft_#{$var * 5} > div > div > div > :nth-child(n + 2) {
    margin-left: #{$var * 5}px;
  }
}
