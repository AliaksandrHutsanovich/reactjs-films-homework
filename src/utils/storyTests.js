const storyTests = (stories, renderer) => {
  Object.keys(stories).forEach((storyKey) => {
    if (typeof stories[storyKey] !== 'function') {
      return;
    }
    it(`${storyKey} should be rendered properly`, () => {
      renderer.render(stories[storyKey]());
      const result = renderer.getRenderOutput();

      expect(result).toMatchSnapshot();
    });
  });
};

export default storyTests;
