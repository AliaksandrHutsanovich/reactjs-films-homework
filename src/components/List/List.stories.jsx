import React from 'react';
import List from './List';

export default {
  component: List,
  title: 'components/List',
  decorators: [
    (story) => (
      <div
        style={{
          backgroundColor: 'red',
          padding: '20px',
          width: '800px',
          height: '400px',
        }}
      >
        {story()}
      </div>
    ),
  ],
};

const names = [
  { id: 1, name: 'Alex' },
  { id: 2, name: 'Valdis' },
  { id: 3, name: 'Gustav' },
  { id: 4, name: 'Vlad' },
];

export const verticalList = () => (
  <List
    dataSource={names}
    size="small"
  />
);

export const horizontalList = () => (
  <List
    dataSource={names}
    size="small"
    grid={{ gutter: 12, column: 4 }}
    itemAutoWidth
  />
);

export const emptyList = () => (
  <List
    dataSource={[]}
    size="large"
  />
);
