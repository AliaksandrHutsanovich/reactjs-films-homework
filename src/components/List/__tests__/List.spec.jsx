import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import storyTests from '../../../utils/storyTests';
import * as stories from '../List.stories';
import { renderListItem } from '../List';

describe('<List />', () => {
  const renderer = new ShallowRenderer();

  storyTests(stories, renderer);

  it('list item should be rendered properly', () => {
    const ListItem = ({ item }) => renderListItem(item);
    renderer.render(<ListItem item={{ id: 0, text: 'retop' }} />);
    const result = renderer.getRenderOutput();

    expect(result).toMatchSnapshot();
  });
});
