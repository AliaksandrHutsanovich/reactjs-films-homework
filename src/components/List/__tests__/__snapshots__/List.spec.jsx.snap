// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> emptyList should be rendered properly 1`] = `
<List
  className=""
  dataSource={Array []}
  size="large"
/>
`;

exports[`<List /> horizontalList should be rendered properly 1`] = `
<List
  className="list"
  dataSource={
    Array [
      Object {
        "id": 1,
        "name": "Alex",
      },
      Object {
        "id": 2,
        "name": "Valdis",
      },
      Object {
        "id": 3,
        "name": "Gustav",
      },
      Object {
        "id": 4,
        "name": "Vlad",
      },
    ]
  }
  grid={
    Object {
      "column": 4,
      "gutter": 12,
    }
  }
  size="small"
/>
`;

exports[`<List /> list item should be rendered properly 1`] = `
<Item
  className="listItem"
>
  <UNDEFINED
    size={12}
    type="white"
  />
</Item>
`;

exports[`<List /> verticalList should be rendered properly 1`] = `
<List
  className=""
  dataSource={
    Array [
      Object {
        "id": 1,
        "name": "Alex",
      },
      Object {
        "id": 2,
        "name": "Valdis",
      },
      Object {
        "id": 3,
        "name": "Gustav",
      },
      Object {
        "id": 4,
        "name": "Vlad",
      },
    ]
  }
  size="small"
/>
`;
